File: engine/src/Engine.cpp
#include "Engine.h"

Engine::Engine(const EngineConfig& config_, std::shared_ptr<DataHandler> data_handler_) {
    config = config_;
    data_handler = data_handler_;

    execution_handler = std::make_unique<ExecutionHandler>();
    context = Context{this};

    execution_handler->set_context(&context);

    if (data_handler) {
        data_handler->set_context(&context);
    }
}

Engine::~Engine() = default;

void Engine::publish(std::unique_ptr<Event> event) {
    // Push incoming events (Order/Fill/Signal) onto the engine queue for processing
    event_queue.push_back(std::move(event));
}

void Engine::set_config(const EngineConfig& config_) {
    config = config_;
}

void Engine::set_data_handler(std::shared_ptr<DataHandler> data_handler_) {
    data_handler = data_handler_;

    if (data_handler) {
        data_handler->set_context(&context);
    }
}

void Engine::set_strategy(std::shared_ptr<Strategy> strategy_) {
    strategy = strategy_;

    if (strategy) {
        strategy->set_context(&context);
    }
}

void Engine::set_portfolio(std::shared_ptr<Portfolio> portfolio_) {
    portfolio = portfolio_;

    if (portfolio) {
        portfolio->set_context(&context);
    }
}

void Engine::run_backtest() {
    // 1. Initialize strategy
    strategy->on_start();

    // 2. Start backtest
    while (data_handler->has_next()) {
        // Queue should be empty at this point
        event_queue.push_back(data_handler->next());

        // 4. Now iterate through event queue
        while (!event_queue.empty()) {
            std::unique_ptr<Event> event = std::move(event_queue.front());
            event_queue.pop_front();

            switch (event->type) {
                case EventType::MARKET: {
                    MarketEvent& market_event = dynamic_cast<MarketEvent&>(*event);

                    execution_handler->on_market_update(market_event);
                    strategy->on_update(market_event);
                    break;
                }

                case EventType::SIGNAL: {
                    SignalEvent& signal_event = dynamic_cast<SignalEvent&>(*event);
                    portfolio->on_signal(*signal_event.signal);
                    break;
                }

                case EventType::ORDER: {
                    OrderEvent& order_event = dynamic_cast<OrderEvent&>(*event);
                    execution_handler->submit_order(std::move(order_event.order));
                    break;
                }

                case EventType::FILL: {
                    FillEvent& fill_event = dynamic_cast<FillEvent&>(*event);
                    portfolio->on_fill(*fill_event.fill);
                    break;
                }
            }
        }
    }
}

File: src/Engine.py
from dotenv import load_dotenv
import os

from src.Alert import send_alert

from alpaca.trading.client import TradingClient
from alpaca.trading.stream import TradingStream

from apscheduler.schedulers.background import BackgroundScheduler
from pytz import timezone

from src.Strategy import Strategy
from src.Portfolio import Portfolio
from src.ExecutionHandler import ExecutionHandler
from src.Context import Context, EventSink
from db import create_fill, create_order
from src.Events import *
from src.Types import *

load_dotenv()

ALPACA_API_KEY = os.getenv("ALPACA_API_KEY")
ALPACA_SECRET = os.getenv("ALPACA_SECRET")

class Engine(EventSink):
    def __init__(self):
        self.strategy: Strategy = None
        self.portfolio: Portfolio = None

        self.event_queue: list[Event] = []

        self.trading_client: TradingClient = TradingClient(ALPACA_API_KEY, ALPACA_SECRET, paper=True)
        self.trading_stream: TradingStream = TradingStream(ALPACA_API_KEY, ALPACA_SECRET, paper=True)

        self.trading_stream.subscribe_trade_updates(self.handle_trading_stream_updates)

        self.execution_handler: ExecutionHandler = ExecutionHandler(self.trading_client)

        self.scheduler = BackgroundScheduler()
        self.market_tz = timezone("America/New_York")

    def schedule_tasks(self):
        self.scheduler.add_job(
            self.generate_market_open_event,
            trigger="cron",
            day_of_week="mon-fri",
            hour=9,
            minute=30,
            timezone=self.market_tz,
            id="market_open_event"
        )

        self.scheduler.start()

    def generate_market_open_event(self):
        event = MarketEvent()

        self.handle_update(event)

    def publish(self, event: Event):
        self.event_queue.append(event)

    def set_strategy(self, strategy: Strategy):
        self.strategy = strategy
        self.strategy.set_context(Context(event_sink=self))

    def set_portfolio(self, portfolio: Portfolio):
        self.portfolio = portfolio
        self.portfolio.set_context(Context(event_sink=self))

    async def handle_trading_stream_updates(self, data):
        try:
            if data.event == "new":
                create_order(order_id=data.order.id, symbol=data.order.symbol, quantity_ordered=float(data.order.qty), status="pending")

                send_alert(f"New order event received from trading stream. \n {data.order.symbol} {data.order.qty} @ {data.order.limit_price if data.order.limit_price else 'MKT'}")

            elif data.event == "fill" or data.event == "partial_fill":
                event = FillEvent(
                    timestamp=data.timestamp,
                    fill=Fill(
                        symbol=data.order.symbol,
                        quantity=float(data.qty),
                        side=data.order.side.upper(),
                        fill_price=float(data.price),
                        commission=0.0 # Alpaca does not provide commission data
                    )
                )

                # TODO: Move to portfolio
                create_fill(order_id=data.order.id, quantity=float(data.qty), price=float(data.price), filled_at=data.timestamp)

                send_alert(f"New fill event received from trading stream. \n {data.order.symbol} {data.qty} @ {data.price}")

                self.handle_update(event)

        except Exception as e:
            send_alert(f"Error processing trading stream update: {str(e)}")

    def handle_update(self, event: Event):
        # Push event to event queue
        self.event_queue.append(event)

        # Run until event queue is empty
        while self.event_queue:
            current_event = self.event_queue.pop(0)

            if current_event.type == EventType.MARKET:
                self.strategy.on_update(current_event)
            elif current_event.type == EventType.SIGNAL:
                self.portfolio.on_signal(current_event)
            elif current_event.type == EventType.ORDER:
                self.execution_handler.execute_order(current_event)
            elif current_event.type == EventType.FILL:
                self.portfolio.on_fill(current_event)

    def run(self):
        self.schedule_tasks()
        self.trading_stream.run()